File: Sources/ReixKernel/Core/Stubs.c
//
//  Stubs.c
//  ReixOS
//
//  Created by Eliomar Alejandro Rodriguez Ferrer on 21/04/2026.
//

#include <stdint.h>
#include <stddef.h>

// --- Protezione Stack ---
// Un valore casuale per il canarino (puoi cambiarlo all'avvio per sicurezza)
uintptr_t __stack_chk_guard = 0x595e9fbd394d2c87;

// Funzione chiamata se lo stack è corrotto
void __stack_chk_fail(void) {
    // In un kernel serio, qui chiameresti un 'panic'
    while (1) {
        __asm__ volatile("wfi");
    }
}

// --- Gestione Memoria (Temporanea) ---
void free(void *ptr) {
    // Per ora non facciamo nulla
}

int posix_memalign(void **memptr, size_t alignment, size_t size) {
    // Inizialmente puoi restituire un errore (es. ENOMEM)
    // o implementare un'allocazione statica temporanea.
    return -1;
}

// --- Entropia (Random) ---
void arc4random_buf(void *buf, size_t nbytes) {
    uint8_t *p = (uint8_t *)buf;
    for (size_t i = 0; i < nbytes; i++) {
        p[i] = 0x42; // Un valore "casuale" molto prevedibile per ora
    }
}

File: Sources/ReixKernel/Core/Stubs.swift
//
//  Stubs.swift
//  ReixOS
//
//  Freestanding runtime symbols the toolchain expects but that the kernel does
//  not really use: the stack-protector canary/handler and the libc allocation
//  and entropy hooks. Pure-Swift replacements for the former `Stubs.c`.
//

@_silgen_name("__stack_chk_guard")
public var __stack_chk_guard: UInt = 0x595e9fbd394d2c87

@_cdecl("__stack_chk_fail")
public func __stack_chk_fail() {
    while true {}
}

@_cdecl("free")
public func free(_ ptr: UnsafeMutableRawPointer?) {}

@_cdecl("posix_memalign")
public func posix_memalign(
    _ memptr   : UnsafeMutablePointer<UnsafeMutableRawPointer?>?,
    _ alignment: Int,
    _ size     : Int
) -> Int32 {
    -1
}

@_cdecl("arc4random_buf")
public func arc4random_buf(_ buf: UnsafeMutableRawPointer?, _ nbytes: Int) {
    guard let p = buf?.assumingMemoryBound(to: UInt8.self) else { return }
    for i in 0..<nbytes { p[i] = 0x42 }
}
